### PR TITLE
[k8s] Attach custom metadata to objects created by SkyPilot

### LINKED ISSUE #3317: [k8s] Add ability to add custom annotations/metadata to k8s objects created by SkyPilot.
Requested by user:
> Is it possible to add annotations to the kubernetes service accounts created by sky?
> ...
> The annotations im looking to add would only be relevant for the service account, but if it’s easier to add to everything I don’t think it’d cause issues

### PR DESCRIPTION
**Pull request by romilbhardwaj**
Closes #3317. Adds the ability to attach custom metadata to all objects created by SkyPilot. Useful for accounting and management of SkyPilot specific resources.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Manual testing by inspecting resources with `kubectl describe ...` with both `ingress` and `loadbalancer` port modes.
- [x] Kubernetes smoke tests in progress `pytest tests/test_smoke.py --kubernetes`

### CODE REVIEW

File: docs/source/reference/config.rst [hunk 1]
```diff
@@ -276,6 +276,18 @@ Available fields and semantics:
...
+    #
+    # Uses the same schema as Kubernetes metadata object: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#objectmeta-v1-meta
+    #
+    # Since metadata is applied to all all objects created by SkyPilot,
+    # specifying 'name' and 'namespace' fields here is not allowed.
+    custom_metadata:
+      labels:
+        mylabel: myvalue
+      annotations:
+        myannotation: myvalue
```
Comment (Michaelvll) [inline docs/source/reference/config.rst:289]: Are both of the `labels` and `annotations` required? We are using `instance_tags:` for the other clouds to allow user specify custom tags for all the VMs created. Would it be possible to have a similar argument for k8s pods?

**Review by Michaelvll — COMMENTED** [review]
Thanks for the quick fix @romilbhardwaj! Left a question : )

File: docs/source/reference/config.rst (hunk 1 above)
Comment (romilbhardwaj) [reply docs/source/reference/config.rst:289]: Both are not required, they can be different or even just one can be present. Labels and annotations serve different purposes - labels are used for filtering, selecting and scheduling, while annotations are metadata that can be consumed by third party tools (e.g., `prometheus.io/scrape` annotations tells prometheus which objects to scrape). 

Given this, users may want to set labels and/or annotations, this config gives them the flexibility to do so.

File: sky/provision/kubernetes/utils.py [hunk 2]
```diff
@@ -812,6 +812,10 @@ def setup_ssh_jump_svc(ssh_jump_name: str, namespace: str,
     # Fill in template - ssh_key_secret and ssh_jump_image are not required for
     # the service spec, so we pass in empty strs.
     content = fill_ssh_jump_template('', '', ssh_jump_name, service_type.value)
+
+    # Add custom metadata from config
+    merge_custom_metadata(content['service_spec']['metadata'])
```
Comment (Michaelvll) [inline sky/provision/kubernetes/utils.py:817]: Q: If we have multiple users sharing the same jump pod, and they have different metadata field setup in their `~/.sky/config.yaml`. What will happen?

File: sky/provision/kubernetes/utils.py [hunk 3]
```diff
@@ -1059,7 +1068,48 @@ def get_endpoint_debug_message() -> str:
                                           debug_cmd=debug_cmd)
 
 
-def combine_pod_config_fields(config_yaml_path: str) -> None:
+def merge_dicts(source: Dict[Any, Any], destination: Dict[Any, Any]):
```
Comment (Michaelvll) [inline sky/provision/kubernetes/utils.py:1071]: make this function private? Seems only used within the same file?

File: docs/source/reference/config.rst (hunk 1 above)
Comment (Michaelvll) [reply docs/source/reference/config.rst:289]: Makes sense! Maybe in the future, we can provide also provide `instance_tags` as an alias for specifying both `labels` and `annotations` with the same set of key, values.

**Review by Michaelvll — APPROVED** [review]
Thanks for adding this support @romilbhardwaj! LGTM.

File: sky/provision/kubernetes/utils.py (hunk 3 above)
Comment (romilbhardwaj) [reply sky/provision/kubernetes/utils.py:1071]: It's also used in authentication.py, where we use merge_dicts when setting up the secrets..

File: sky/provision/kubernetes/utils.py (hunk 2 above)
Comment (romilbhardwaj) [reply sky/provision/kubernetes/utils.py:817]: Great question - the metadata specified during the first creation of the ssh jump pod will be used throughout it's life. Other users sharing the jump pod will not have their metadata overwrite the existing pod and service's metadata.

**Pull request merged**